contracts/HeapOrdering.sol

```diff
@@ -0,0 +1,318 @@
+    /// @param _heap The heap to modify.
+    /// @param _id The address of the account to decrease the amount.
+    /// @param _newValue The new value of the account.
+    function decrease(
+        HeapArray storage _heap,
+        address _id,
+        uint256 _newValue
+    ) private {
+        uint256 rank = _heap.ranks[_id];
+        setAccountValue(_heap, rank, _newValue);
+
+        // We only need to restore the invariant if the account is a node in the heap
+        if (rank <= _heap.size >> 1) shiftDown(_heap, rank);
+    }
+
+    /// @notice Increases the amount of an account in the `_heap`.
+    /// @dev Only call this function when `_id` is in the `_heap` with a smaller value than `_newValue`.
+    /// @param _heap The heap to modify.
+    /// @param _id The address of the account to increase the amount.
+    /// @param _newValue The new value of the account.
+    /// @param _maxSortedUsers The maximum size of the heap.
+    function increase(
+        HeapArray storage _heap,
+        address _id,
+        uint256 _newValue,
+        uint256 _maxSortedUsers
+    ) private {
+        uint256 rank = _heap.ranks[_id];
+        Account memory account = getAccount(_heap, rank);
+        account.value = _newValue;
+        setAccountValue(_heap, rank, _newValue);
+        uint256 nextSize = _heap.size + 1;
+
+        if (rank < nextSize) shiftUp(_heap, rank);
+        else if (nextSize <= _heap.accounts.length) {
+            swap(_heap, nextSize, rank);
+            shiftUp(_heap, nextSize);
+            _heap.size = computeSize(nextSize, _maxSortedUsers);
+        }
+    }
+
+    /// @notice Removes an account in the `_heap`.
+    /// @dev Only call when this function `_id` is in the `_heap` with value `_removedValue`.
+    /// @param _heap The heap to modify.
+    /// @param _id The address of the account to remove.
+    /// @param _removedValue The value of the account to remove.
+    function remove(
+        HeapArray storage _heap,
+        address _id,
+        uint256 _removedValue
+    ) private {
+        uint256 rank = _heap.ranks[_id];
+        uint256 accountsLength = _heap.accounts.length;
+
+        // Swap the last account and the account to remove, then pop it.
+        swap(_heap, rank, accountsLength);
+        if (_heap.size == accountsLength) _heap.size--;
+        _heap.accounts.pop();
+        delete _heap.ranks[_id];
+
+        // If the swapped account is in the heap, restore the invariant: its value can be smaller or larger than the removed value.
+        if (rank <= _heap.size) {
+            if (_removedValue > getAccount(_heap, rank).value) shiftDown(_heap, rank);
+            else shiftUp(_heap, rank);
+        }
+    }
+
+    /// GETTERS ///
+
+    /// @notice Returns the number of users in the `_heap`.
+    /// @param _heap The heap parameter.
+    /// @return The length of the heap.
+    function length(HeapArray storage _heap) internal view returns (uint256) {
```

Comment:
why difference where between private and internal?

Reply:
It's exactly the same in terms of gas consumption (I tested it to be sure), but the private functions are not designed to be used directly by other contracts using this data-structure. For example, the `remove` function assumes the given address corresponds to an account in the heap; it does not check that this is the case with an assert. This is sound because we only use the remove function in our context where we know that those assumptions hold